### ISSUE TITLE
ASSIST XDECI behaviour not correct

### ISSUE THREAD

**Issue by adelosa** [z390development/z390#291]
In reviewing the documentation for z390 ASSIST, the function was compared to the original ASSIST assembler doco.

> XDECI is generally used to scan a data card read by XREAD. The sequence of actions performed by XDECI is as follows:
> 1. Beginning at the location given by ADDRESS, memory is scanned for the first character which is not a blank.
> 2. If the first character found is anything but a decimal digit or plus or minus sign, register 1 is set to the address of that character, and the condition code is set to 3 (overflow) to show that no decimal number could be converted. The contents of REG are not changed, and nothing more is done.
> 3. From one to nine decimal digits are scanned, and the number converted to binary and placed in REG, with the appropriate sign. The condition code is set to 0 (0), 1 (-), or 2 (+), depending on the value just placed in REG.
> 4. Register 1 is set to the address of the first non-digit after the string of decimal digits. Thus REG should not usually be 1. This permits the user to scan across a card image for any number of decimal values. The values should be separated by blanks, since otherwise the scanner could hang up on a string like -123*, unless the user checks for this himself. I.e. XDECI will skip leading blanks but will not itself skip over any other characters.
> 5. During step 3, if ten or more decimal digits are found, register 1 is set to the address of the first character found which is not a decimal digit, the condition code is set to 3, and REG is left unchanged. A plus or minus sign alone causes a similar action, with R1 set to the address of the character following the sign character.

* Item 3 is not aligned. The z390 implementation does not set the condition code based on the sign of the result. If converted, it always sets the condition code to zero.
* Item 5 is not aligned. The z390 implementation will continue to convert after 9 digits and does not set CC=3.

I wrote the following program to test the behaviour as described above. Obviously the ASSIST option needs to be enabled to run this program.

My view is the following program should run successfully if the XDECI instruction has been implemented correctly.
**NOTE**: I am a beginner assembler programmer and this is probably not how an experienced programmer would create the program but its what I know. Any ideas on improvements would also be appreciated. 

The fix should include the test program as part of the regression test suite.
 
```
* 
* ASSIST1 - Tests ASSIST XDECI function
*
ASSIST1  SUBENTRY
         USING TESTDATA,6      R6 to use testdata dsect
         LA    7,NUMTESTS      R7 contains test counter
         LA    6,TEST1         R6 points at current testdata
LOOP     PERFORM CHKXDECI      Perform the test
         LA    6,TDATALEN(,6)  Increment R6 to next test
         BCT   7,LOOP          Loop until all tests complete
         SUBEXIT

CHKXDECI PENTRY
* R5=XDECI output
         MVC   PRINTLN(3),=C'***'
         MVC   PRINTLN+4(20),TESTDESC
         XPRNT PRINTLN,80
         SR    1,1             Clear R1
         SR    5,5             Clear R5
         XDECI 5,INPUTVAL      Perform XDECI instruction
*
* check cc
*
         BRC   B'1000',CC0
         BRC   B'0100',CC1
         BRC   B'0010',CC2
         BRC   B'0001',CC3
         ABEND 8               No cc set
CC0      CLI   EXPCCVAL,C'0'   Is expected CC=0
         J     CHKCC
CC1      CLI   EXPCCVAL,C'1'   Is expected CC=1
         J     CHKCC
CC2      CLI   EXPCCVAL,C'2'   Is expected CC=2
         J     CHKCC
CC3      CLI   EXPCCVAL,C'3'   Is expected CC=3
CHKCC    JNE   CCERR           Not expected CC
         XPRNT =C'CC OK',5
         J     CHKOUT
CCERR    XPRNT =C'CC not OK',9
         ABEND 901           
*
* check output
*
CHKOUT   C     5,EXPOUT        Compare output to expected
         JE    OUTOK           OK if equal
         J     OUTNOTOK        Otherwise it's a test issue
OUTOK    XPRNT =C'Output OK',9
         J     CHKR1
OUTNOTOK XPRNT =C'Output ERR',10
         ABEND 900
*
* check r1 value
*
CHKR1    C     1,EXPGR1
         JE    R1EQU
         J     R1NOTEQU
R1EQU    XPRNT =C'R1 OK',5
         J     ECHK
R1NOTEQU XPRNT =C'R1 not OK',9
ECHK     PEXIT

*
* Test conditions table - specify input and expected output
*
TESTTBL  DS    0F
TEST1    DC    F'0'          Exp output
         DC    A(TEST1IN+1)  Exp GR1
         DC    C'0'          Exp CC
TEST1IN  DC    CL12'0 '      Input
         DC    CL20'Check "0"'

TEST2    DC    F'1234'       Exp output
         DC    A(TEST2IN+5)  Exp GR1
         DC    C'2'          Exp CC
TEST2IN  DC    CL12'1234+ '  Input  
         DC    CL20'Check "1234+"'

TEST2A   DC    F'-1234'      Exp output
         DC    A(TEST2AIN+5) Exp GR1
         DC    C'1'          Exp CC
TEST2AIN DC    CL12'1234- '  Input 
         DC    CL20'Check "1234-"'

TEST3    DC    F'0'          Exp output
         DC    F'0'          Exp GR1
         DC    C'3'          Exp CC
         DC    CL12' X'      Input
         DC    CL20'Check no valid chars'

TEST4    DC    F'123456789'  Exp output
         DC    A(TEST4IN+9)  Exp GR1
         DC    C'2'          Exp CC
TEST4IN  DC    CL12'123456789 '  Input
         DC    CL20'Check max chars (9)'

TEST5    DC    F'0'          Exp output
         DC    F'0'          Exp GR1
         DC    C'3'          Exp CC
         DC    CL12'1234567890 '  Input
         DC    CL20'Check 10 chars ERR'

         DS    0F
NUMTESTS EQU   (*-TESTTBL)/TDATALEN
         DC    A(NUMTESTS)   
PRINTLN  DC    CL80' ' 

TESTDATA DSECT ,
EXPOUT   DS    F
EXPGR1   DS    F
EXPCCVAL DS    CL1
INPUTVAL DS    CL12
TESTDESC DS    CL20
         DS    0F
TDATALEN EQU   *-TESTDATA
         END
```

**Comment by abekornelis** [comment]
John Ganci wrote: 
ASSIST is EBCDIC only. The z390 XDECI code in pz390.java handles ASCII or EBCDIC. The z390 XHEXI code in pz390.java handles EBCDIC only.

Question: Should the z390 code adhere to ASSIST (EBCDIC only) or should it allow both ASCII and EBCDIC digits (decimal for XDECI, hexadecimal for XHEXI)?

I think EBCDIC only.

Regards,
John Ganci

**Comment by abekornelis** [comment]
@jyganci 

I think that z390 should primarily follow ASSIST definitions.

However, we have defined the allow/noallow option to enable/disable
various z390 extensions that may not operate compatibly with IBM.

To my mind, when allow is turned on, any input codepage
should be eligible for proper processing.
Maybe there are other options relating to code page
that might also influence operation of these instructions?

You might want to double-check on that ...

**Comment by adelosa** [comment]
Fixed by PR #331
